### PR TITLE
landscaper/controlplane 6: Add landscaper resources

### PR DESCRIPTION
**Pull request by danielfoehrKn**
**How to categorize this PR?**

/kind enhancement

**What this PR does / why we need it**:

Adds the blueprint for the controlplane component
- automatically generated based on the golang types
- reusable for other components (such as the gardenlet)
- adds the blueprint to the component descriptor
**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
- based on #5130  => though it can be reviewed independently 

I successfully rendered the `deployItems` for the blueprint locally. That should have verified the import configuration against the JSONSchema in the blueprints import configuration.

Please review the commits (the other commits are from the previous PR #5130 and will be rebased once this PR is merged)
- remove unused fields from the controlplane config
- move the gardenlet blueprint to a dedicated directory to reduce the CTF size
- add generated blueprint files
- add generator code
- add the controlplane blueprint to the g/g component descriptor

**Release note**:

```other operator
Added the blueprint for the landscaper-controlplane component
```
```other operator
Added a blueprint and OpenAPI generator for landscaper components
```

### CODE REVIEW

**Comment by achimweigel** [comment]
Very interesting approach to generate the json schema from the definitions and then the blueprint with all these definitions. This really helps with components with such complex import data. Would it be possible to format the json files in the blueprint directory for better readability?

**Comment by danielfoehrKn** [comment]
> Would it be possible to format the json files in the blueprint directory for better readability?

What formatting would you like to see? 
I suppose so that it is not a one liner?
I could also format it as a `yaml`.

**Comment by achimweigel** [comment]
Yes not just a one liner, just some kind of json pretty print - nothing fancy.

File: landscaper/pkg/controlplane/generate/generate-openapi.sh
Comment (achimweigel) [inline landscaper/pkg/controlplane/generate/generate-openapi.sh]: Do we really need to maintain all dependencies here? And what is about indirect dependencies?

Comment (danielfoehrKn) [reply landscaper/pkg/controlplane/generate/generate-openapi.sh]: Yes, you need to cover all the types which are part of your import definition. Otherwise, JSONSchemas / OpenAPI for dependencies is missing. For missing dependencies, an "ALLOW ALL" JSONSchema is inserted.

Comment (danielfoehrKn) [reply landscaper/pkg/controlplane/generate/generate-openapi.sh]: > And what is about indirect dependencies?

They are not automatically covered (need to explicitly be included)

**Comment by danielfoehrKn** [comment]
> Yes not just a one liner, just some kind of json pretty print - nothing fancy.

The blueprints JSONSchema is now rendered as yaml - I think that makes it much more readable. Originally I thought no human will ever read that manually, but I guess you never know :) 
Thanks for the suggestion.

File: landscaper/pkg/controlplane/generate/generate-openapi.sh
Comment (danielfoehrKn) [reply landscaper/pkg/controlplane/generate/generate-openapi.sh]: Just FYI: Even though the generated OpenAPI-go code can return the OpenAPI for all the types in the specified input directories, It does not mean the Import configuration of the landscaper controlplane uses all of these types. 
For instance, the directory `k8s.io/api/core/v1` contains a specific type that we need, but not all the types in the directory are used in the Import configuration of the landscaper component (the OpenAPIGenerator only works on file, not type level).

Only for types that are referenced and used in the Import configuration of the landscaper controlplane, the JSON Schema is generated for the blueprint.

Comment (achimweigel) [reply landscaper/pkg/controlplane/generate/generate-openapi.sh]: OK. The solution would be perfect, if you would not need to add all dependencies but only the import folder. Perhaps we could later provide a tool for this, which also computes these dependencies.  
 
> Only for types that are referenced and used in the Import configuration of the landscaper controlplane, the JSON Schema is generated for the blueprint.

I have seen this. I think especially parts of your blueprint generation algorithm would be also interesting for others.

Comment (danielfoehrKn) [reply landscaper/pkg/controlplane/generate/generate-openapi.sh]: @achimweigel I added a re-usable generator for the input directories in the last commit. Would be nice if you can take another look.

The size of the generated go-code doubled from 1 -> 2MB due to all transitive dependencies being included. The end result, the blueprint filesystem remains the same, however.

**Review by achimweigel — DISMISSED** [review]
/lgtm 

I like the methods including automatic dependency detection. I only do not understand why you restrict this to k8s and github packages

**Comment by danielfoehrKn** [comment]
> I only do not understand why you restrict this to k8s and github packages

To reduce the search (no standard libs included). All k8s repositories are under `k8s.io` and all Gardener repos that could contain and `openapi-gen` tags are on github.

**Comment by danielfoehrKn** [comment]
@rfranzke I need at least one core reviewer to be able to merge would be very nice if you could take a quick look

**Comment by rfranzke** [comment]
@danielfoehrKn I can certainly approve (without understanding/reviewing much) since the changes are only related to the `landscaper` dir, however, I saw that there is quite some generation for core Kubernetes types. Is all of this really needed? From my outside perspective, this seems to be too much and incredibly overblown. Does each Landscaper component need to do this generation? We can get this merged, but I fear we are not moving to the right direction here...

**Comment by danielfoehrKn** [comment]
>  saw that there is quite some generation for core Kubernetes types. Is all of this really needed

@rfranzke  Let me give you some background: Landscaper describes each deliverable using a resource called "Blueprint" in a registry. Blueprints need to define their imported and exported fields. The format of how the landscaper describes imports is via JSONSchema. 
The idea is that the landscaper can validate any Installation resource using its associated blueprints JSONSchema.
This is very much comparable to the `openAPIV3Schema` field for [Kubernetes CRDs](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#specifying-a-structural-schema).

Unfortunately, the controlplane component has a large configuration file with many types. This creates a very large JSONSchema document to validate it.

We could opt to not use landscaper's validation options by using "allow all" - type JSONSchema.
 - not having JSONSchema validation would remove a main advantage of the landscaper: to be able to validate an Installation before running it + to precisely point out configuration problems. Not having that makes usage of landscaper components in my point of view even more difficult.
 - there are features planned "on top" - e.g a UI that helps to create the necessary configuration files


In the future, the landscaper team might create a set of pre-defined types for Kubernetes types in the landscaper itself to make the JSONSchema in each component more concise.  This is again similar to Kubernetes CRDs which can implicitly validate fields such as `metadata`.
However, this is not done yet + quite some effort with questionable ROI.

> quite some generation

It is not feasible to create & maintain JSONSchema by hand for such big configuration files.
I see it as the only way how we can reduce the effort for new components and make it maintainable in the long run.

Overall, the big amount of generated JSONSchema is a result of the Landscaper's choice to use JSONSchema for validation.

**Comment by rfranzke** [comment]
Well, as said, I don't want to block this PR, so let's merge it after `v1.39` got cut.

**Comment by danielfoehrKn** [comment]
@rfranzke let's get this in as two subsequent PRs are waiting. Thanks

**Review by rfranzke — APPROVED** [review]

**Pull request merged**
